### PR TITLE
Attempt to add SDL2_Surface / SDL3_Surface

### PR DESCRIPTION
**Pull request by 1bsyl**

### CODE REVIEW

File: src/sdl2_compat.c [hunk 1]
```diff
@@ -343,8 +343,10 @@ static QuirkEntryType quirks[] = {
 
 #ifdef __linux__
 static void OS_GetExeName(char *buf, const unsigned maxpath) {
+    int ret;
     buf[0] = '\0';
-    readlink("/proc/self/exe", buf, maxpath);
+    ret = readlink("/proc/self/exe", buf, maxpath);
+    (void)ret;
```
Comment (sezero) [inline src/sdl2_compat.c:349]: Are you getting `unused-result` warnings? If so, is doing `(void) readlink(.....)` not enough?

File: test/testautomation_platform.c [hunk 2]
```diff
@@ -386,7 +386,7 @@ int platform_testSetErrorEmptyInput(void *arg)
 int platform_testSetErrorInvalidInput(void *arg)
 {
     int result;
-    const char *invalidError = NULL;
+    const char *invalidError = "";
```
Comment (sezero) [inline test/testautomation_platform.c:389]: Neither SDL3 nor SDL3 do this: why is this needed?

File: src/sdl2_compat.c [hunk 3]
```diff
@@ -1905,6 +1907,80 @@ SDL_WarpMouseGlobal(int x, int y)
...
+        return;
+    }
+
+    s3->flags = s2->flags;
+    s3->format = s2->format;
+    s3->w = s2->w;
+    s3->h = s2->h;
+    s3->pitch = s2->pitch;
+    s3->pixels = s2->pixels;
+    // ignore user data?
```
Comment (sezero) [inline src/sdl2_compat.c:1952]: What to do with `userdata` indeed ?

File: src/sdl2_compat.c [hunk 4]
```diff
@@ -1905,6 +1907,80 @@ SDL_WarpMouseGlobal(int x, int y)
...
+    if (!s2 || !s3) {
+        return;
+    }
+    s2->flags = s3->flags;
+    s2->format = s3->format;
+    s2->w = s3->w;
+    s2->h = s3->h;
+    s2->pitch = s3->pitch;
+    s2->pixels = s3->pixels;
+    // ignore user data?
```
Comment (sezero) [inline src/sdl2_compat.c:1971]: ... as above

File: src/sdl2_compat.c [hunk 5]
```diff
@@ -2340,31 +2416,43 @@ SDL_LoadWAV_RW(SDL2_RWops *rwops2, int freesrc, SDL2_AudioSpec *spec2, Uint8 **a
...
     SDL_Surface *retval = NULL;
+    SDL2_Surface *s2 = NULL;
     SDL_RWops *rwops3 = RWops2to3(rwops2);
     if (rwops3) {
         retval = SDL3_LoadBMP_RW(rwops3, freesrc != 0);
         if (!freesrc) {
             SDL3_DestroyRW(rwops3);  /* don't close it because that'll close the SDL2_RWops. */
         }
+
+        s2 = SDL3_malloc(sizeof(*s2));
```
Comment (sezero) [inline src/sdl2_compat.c:2431]: Missing `(SDL2_Surface *)` cast, leading to failures in CI

File: src/sdl2_compat.c [hunk 6]
```diff
@@ -2486,54 +2574,150 @@ SDL_GetWindowGammaRamp(SDL_Window *window, Uint16 *red, Uint16 *blue, Uint16 *gr
...
+    SDL3_zero(s3);
+    Surface2to3_init(src2, &s3);
+
+    t3 = SDL3_ConvertSurfaceFormat(&s3, pixel_format);
+
+    if (!t3) {
+        return NULL;
+    }
+
+    s2 = SDL3_malloc(sizeof(*s2));
```
Comment (sezero) [inline src/sdl2_compat.c:2616]: .. as above (missing cast)

File: src/sdl2_compat.c [hunk 7]
```diff
@@ -2486,54 +2574,150 @@ SDL_GetWindowGammaRamp(SDL_Window *window, Uint16 *red, Uint16 *blue, Uint16 *gr
...
+DECLSPEC SDL2_Surface * SDLCALL
 SDL_CreateRGBSurfaceWithFormatFrom(void *pixels, int width, int height, int depth, int pitch, Uint32 format)
 {
-    return SDL3_CreateSurfaceFrom(pixels, width, height, pitch, format);
+    SDL_Surface *s3;
+    SDL2_Surface *s2;
+
+    s3 = SDL3_CreateSurfaceFrom(pixels, width, height, pitch, format);
+
+    s2 = SDL3_malloc(sizeof(*s2));
```
Comment (sezero) [inline src/sdl2_compat.c:2676]: .. as above (missing cast)

File: src/sdl2_compat.c [hunk 8]
```diff
@@ -7291,6 +7484,347 @@ SDL_AndroidGetExternalStorageState(void)
...
+    SDL_Surface s3;
+    SDL3_zero(s3);
+    Surface2to3_init(s2, &s3);
+    return SDL3_CreateColorCursor(&s3, b, c);
+}
+
+DECLSPEC SDL_Renderer * SDLCALL SDL_CreateSoftwareRenderer(SDL2_Surface *s2)
+{
+    SDL_Surface *s3;
+    s3 = SDL3_malloc(sizeof(*s3));
```
Comment (sezero) [inline src/sdl2_compat.c:7499]: ... as above - missing cast

File: src/sdl2_compat.c [hunk 9]
```diff
@@ -7291,6 +7484,347 @@ SDL_AndroidGetExternalStorageState(void)
...
+
+DECLSPEC int SDLCALL SDL_LockTextureToSurface(SDL_Texture *a, const SDL_Rect *b, SDL2_Surface **s2)
+{
+    int ret;
+    SDL_Surface *s3 = NULL;
+    SDL2_Surface *t2 = NULL;
+
+    ret = SDL3_LockTextureToSurface(a, b, &s3);
+
+    t2 = SDL3_malloc(sizeof(*t2));
```
Comment (sezero) [inline src/sdl2_compat.c:7734]: .. as above  - missing cast

File: src/sdl2_compat.c (hunk 1 above)
Comment (1bsyl) [reply src/sdl2_compat.c:349]:
```
sdl2_compat.c:347:11: error: ignoring return value of ‘readlink’ declared with attribute ‘warn_unused_result’ [-Werror=unused-result]
  347 |     (void)readlink("/proc/self/exe", buf, maxpath);

```

I had to patch this way :(
(this should be part of this PR)

File: src/sdl2_compat.c [hunk 10]
```diff
@@ -7291,6 +7484,347 @@ SDL_AndroidGetExternalStorageState(void)
...
+    Surface3to2_init(&t3, t2);
+    return ret;
+}
+
+
+DECLSPEC int SDLCALL SDL_LockTextureToSurface(SDL_Texture *a, const SDL_Rect *b, SDL2_Surface **s2)
+{
+    int ret;
+    SDL_Surface *s3 = NULL;
+    SDL2_Surface *t2 = NULL;
```
Comment (sezero) [inline src/sdl2_compat.c:7730]: unnecessary initializer

File: src/sdl2_compat.c [hunk 11]
```diff
@@ -7291,6 +7484,347 @@ SDL_AndroidGetExternalStorageState(void)
...
+}
+
+
+DECLSPEC void SDLCALL SDL_FreeSurface(SDL2_Surface *s2)
+{
+    SDL_Surface *s3;
+    if (!s2) {
+        return;
+    }
+    s3 = SDL3_malloc(sizeof(*s3));
```
Comment (sezero) [inline src/sdl2_compat.c:7751]: .. as above (missing cast)

File: test/testautomation_platform.c (hunk 2 above)
Comment (sezero) [reply test/testautomation_platform.c:389]: If this is really needed, it should be done in SDL2 and SDL3 first, and done here as a separate patch

Comment (1bsyl) [reply test/testautomation_platform.c:389]:
```
test/testautomation_platform.c: In function ‘platform_testSetErrorEmptyInput’:
test/testautomation_platform.c:351:28: error: ‘%s’ directive argument is null [-Werror=format-overflow=]
  351 |     result = SDL_SetError("%s", testError);
      |                            ^~
test/testautomation_platform.c:364:64: error: ‘%s’ directive argument is null [-Werror=format-overflow=]
  364 |                             "SDL_GetError(): expected message '%s', was message: '%s'",
```

was just fixing compilation also for me
 (this should not be part of this PR)

File: src/sdl2_compat.c (hunk 1 above)
Comment (sezero) [reply src/sdl2_compat.c:349]: In any case: if this is really needed, it should be done in a separate patch?

File: src/sdl2_compat.c (hunk 3 above)
Comment (1bsyl) [reply src/sdl2_compat.c:1952]: no idea yet :/
that was some experiment

File: test/testautomation_platform.c (hunk 2 above)
Comment (1bsyl) [reply test/testautomation_platform.c:389]: revert, done in other pr

File: src/sdl2_compat.c [hunk 12]
```diff
@@ -1905,6 +1907,80 @@ SDL_WarpMouseGlobal(int x, int y)
...
+    s2->pixels = s3->pixels;
+    // ignore user data?
+    s2->locked = s3->locked;
+    s2->list_blitmap = s3->list_blitmap;
+    s2->clip_rect = s3->clip_rect;
+    s2->map = s3->map;
+    s2->refcount = s3->refcount;
+
+
+
```
[the remark this answers is not in the document]
Comment (1bsyl) [reply src/sdl2_compat.c]: done

File: src/sdl2_compat.c (hunk 6 above)
Comment (1bsyl) [reply src/sdl2_compat.c]: done

File: src/sdl2_compat.c (hunk 3 above)
Comment (1bsyl) [reply src/sdl2_compat.c:1950]: actually, this is ok, in some case.
we don't need the user_data to create a sdl3 surface. we don't know what is the user_data anyway!



but there are other problems see the message below ...

File: src/sdl2_compat.c (hunk 4 above)
Comment (1bsyl) [reply src/sdl2_compat.c:1969]: probably also ok. we don't want to overwrite the user data.
we just want to fetch back the sdl3 surface fields.

**Comment by slouken** [comment]
Hold off on this, I have an idea on how to make the surfaces compatible.

**Comment by sezero** [comment]
Is this still relevant after recent changes?

**Comment by slouken** [comment]
Oh, this is already done, thanks!
